File: cpp-core/src/scoring.cpp
#include "scoring.hpp"

#include <algorithm>
#include <cmath>
#include <limits>

namespace krot {

std::optional<ProfileScore> score_profiles(const std::vector<double>& measured, const std::vector<double>& reference) {
  if (measured.empty() || measured.size() != reference.size()) {
    return std::nullopt;
  }

  const double count = static_cast<double>(measured.size());
  double measured_sum = 0.0;
  double reference_sum = 0.0;
  for (std::size_t index = 0; index < measured.size(); ++index) {
    measured_sum += measured[index];
    reference_sum += reference[index];
  }

  const double measured_mean = measured_sum / count;
  const double reference_mean = reference_sum / count;
  double measured_variance = 0.0;
  double reference_variance = 0.0;
  double covariance = 0.0;
  double error_sq = 0.0;

  for (std::size_t index = 0; index < measured.size(); ++index) {
    const double measured_delta = measured[index] - measured_mean;
    const double reference_delta = reference[index] - reference_mean;
    measured_variance += measured_delta * measured_delta;
    reference_variance += reference_delta * reference_delta;
    covariance += measured_delta * reference_delta;
    const double error = measured[index] - reference[index];
    error_sq += error * error;
  }

  const double denominator = std::sqrt(std::max(0.0, measured_variance * reference_variance));

  // Показатель совпадения показывает, насколько линия измерений похожа на линию с карты.
  return ProfileScore{
      denominator > 0.0 ? covariance / denominator : 0.0,
      std::sqrt(error_sq / count),
  };
}

double relief_m(const std::vector<double>& profile) {
  if (profile.empty()) {
    return 0.0;
  }

  double min_value = std::numeric_limits<double>::infinity();
  double max_value = -std::numeric_limits<double>::infinity();
  for (const double value : profile) {
    min_value = std::min(min_value, value);
    max_value = std::max(max_value, value);
  }
  return max_value - min_value;
}

double standard_deviation_m(const std::vector<double>& profile) {
  if (profile.empty()) {
    return 0.0;
  }

  double sum = 0.0;
  for (const double value : profile) {
    sum += value;
  }
  const double mean = sum / static_cast<double>(profile.size());

  double variance = 0.0;
  for (const double value : profile) {
    variance += std::pow(value - mean, 2.0);
  }
  return std::sqrt(variance / static_cast<double>(profile.size()));
}

}  // namespace krot

File: rust-core/src/scoring.rs
#[derive(Debug, Clone, Copy)]
pub struct ProfileScore {
    pub match_score: f64,
    pub profile_error_m: f64,
}

pub fn score_profiles(measured: &[f64], reference: &[f64]) -> Option<ProfileScore> {
    if measured.is_empty() || measured.len() != reference.len() {
        return None;
    }

    let count = measured.len() as f64;
    let measured_mean = measured.iter().sum::<f64>() / count;
    let reference_mean = reference.iter().sum::<f64>() / count;
    let mut measured_var = 0.0;
    let mut reference_var = 0.0;
    let mut covariance = 0.0;
    let mut error_sq = 0.0;

    for (m, r) in measured.iter().zip(reference.iter()) {
        let dm = *m - measured_mean;
        let dr = *r - reference_mean;
        measured_var += dm * dm;
        reference_var += dr * dr;
        covariance += dm * dr;
        let error = *m - *r;
        error_sq += error * error;
    }

    let denominator = (measured_var * reference_var).max(0.0).sqrt();
    Some(ProfileScore {
        match_score: if denominator > 0.0 {
            covariance / denominator
        } else {
            0.0
        },
        profile_error_m: (error_sq / count).sqrt(),
    })
}

pub(crate) fn relief_m(profile: &[f64]) -> f64 {
    if profile.is_empty() {
        return 0.0;
    }
    let min = profile.iter().copied().fold(f64::INFINITY, f64::min);
    let max = profile.iter().copied().fold(f64::NEG_INFINITY, f64::max);
    max - min
}

pub(crate) fn standard_deviation_m(profile: &[f64]) -> f64 {
    if profile.is_empty() {
        return 0.0;
    }
    let count = profile.len() as f64;
    let mean = profile.iter().sum::<f64>() / count;
    let variance = profile
        .iter()
        .map(|value| (value - mean).powi(2))
        .sum::<f64>()
        / count;
    variance.sqrt()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn identical_profiles_have_high_match() {
        let profile = [300.0, 320.0, 360.0, 330.0, 310.0];
        let score = score_profiles(&profile, &profile).unwrap();
        assert!(score.match_score > 0.999);
        assert!(score.profile_error_m < 0.001);
    }

    #[test]
    fn different_profiles_have_low_match() {
        let measured = [300.0, 320.0, 360.0, 330.0, 310.0];
        let reference = [310.0, 330.0, 300.0, 360.0, 320.0];
        let score = score_profiles(&measured, &reference).unwrap();
        assert!(score.match_score < 0.25);
        assert!(score.profile_error_m > 25.0);
    }
}
